Source/WebGPU/WebGPU/RenderPipeline.mm

```diff
@@ -32,10 +32,258 @@
+}
+
+static MTLPrimitiveType primitiveType(WGPUPrimitiveTopology topology)
+{
+    switch (topology) {
+    case WGPUPrimitiveTopology_PointList:
+        return MTLPrimitiveTypePoint;
+    case WGPUPrimitiveTopology_LineStrip:
+        return MTLPrimitiveTypeLineStrip;
+    case WGPUPrimitiveTopology_TriangleList:
+        return MTLPrimitiveTypeTriangle;
+    case WGPUPrimitiveTopology_LineList:
+        return MTLPrimitiveTypeLine;
+    case WGPUPrimitiveTopology_TriangleStrip:
+        return MTLPrimitiveTypeTriangleStrip;
+    case WGPUPrimitiveTopology_Force32:
+        ASSERT_NOT_REACHED();
+        return MTLPrimitiveTypeTriangle;
+    }
+}
+
+static std::optional<MTLIndexType> indexType(WGPUIndexFormat format)
+{
+    switch (format) {
+    case WGPUIndexFormat_Uint16:
+        return MTLIndexTypeUInt16;
+    case WGPUIndexFormat_Uint32:
+        return MTLIndexTypeUInt32;
+    case WGPUIndexFormat_Undefined:
+        return { };
+    case WGPUIndexFormat_Force32:
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+}
+
+bool Device::validateRenderPipeline(const WGPURenderPipelineDescriptor& descriptor)
+{
+    // FIXME: fully implement this according to the description in
+    // https://gpuweb.github.io/gpuweb/#abstract-opdef-validating-gpurenderpipelinedescriptor
+
+    // FIXME: we additionally reject valid descriptors using unimplemented features.
+    // Remove these checks once we support them.
+
+    // Does not support module constants in vertex shaders
+    if (descriptor.vertex.constantCount)
+        return false;
+    // Does not support vertex shader input buffers
+    if (descriptor.vertex.bufferCount)
+        return false;
+
+    if (descriptor.fragment) {
+        const auto& fragmentDescriptor = *descriptor.fragment;
+
+        // Does not support module constants in fragment shaders
+        if (fragmentDescriptor.constantCount)
+            return false;
+    }
+
+    // Does not support depth stencils.
+    if (descriptor.depthStencil)
+        return false;
+
+    // Does not support multisampling
+    if (descriptor.multisample.count > 1)
+        return false;
+    if (descriptor.multisample.alphaToCoverageEnabled)
+        return false;
+
+    return true;
+}
+
 Ref<RenderPipeline> Device::createRenderPipeline(const WGPURenderPipelineDescriptor& descriptor)
 {
-    UNUSED_PARAM(descriptor);
-    return RenderPipeline::createInvalid(*this);
+    if (!validateRenderPipeline(descriptor))
+        return RenderPipeline::createInvalid(*this);
+
+    MTLRenderPipelineDescriptor* mtlRenderPipelineDescriptor = [MTLRenderPipelineDescriptor new];
+
+    {
+        if (descriptor.vertex.nextInChain)
+            return RenderPipeline::createInvalid(*this);
+
+        const auto& vertexModule = WebGPU::fromAPI(descriptor.vertex.module);
+        const auto& vertexFunctionName = String::fromLatin1(descriptor.vertex.entryPoint);
+
+        const auto vertexFunction = vertexModule.getNamedFunction(vertexFunctionName);
```

Comment:
Missing method `ShaderModule` to grab a handle to a shader function inside the shader library.